baml_language/crates/baml_thir/src/exhaustiveness.rs

```diff
@@ -0,0 +1,833 @@
+//! - **Infinite types**: int, string, classes have infinite value sets
+//!   (only exhaustive via catch-all or type pattern)
+//!
+//! ## Example
+//!
+//! ```baml
+//! enum Status { Active, Inactive, Pending }
+//!
+//! match (s) {
+//!   Status.Active => ...    // ValueSet::EnumVariant("Status", "Active")
+//!   Status.Inactive => ...  // ValueSet::EnumVariant("Status", "Inactive")  
+//!   Status.Pending => ...   // ValueSet::EnumVariant("Status", "Pending")
+//! }
+//! ```
+//!
+//! Each arm covers a single value. Together they cover all values of type Status.
+
+use std::collections::HashMap;
+
+use baml_base::Name;
+use baml_hir::{ExprBody, Literal, MatchArm, Pattern};
+
+use crate::{Db, Ty, lower_type_ref};
+
+// ============================================================================
+// ValueSet: The Core Abstraction
+// ============================================================================
+
+/// Represents a set of runtime values that a pattern can match.
+///
+/// This is the core abstraction for exhaustiveness checking. Unlike types
+/// (which describe what values CAN exist), ValueSet describes what values
+/// a pattern WILL match at runtime.
+///
+/// # Conceptual Model
+///
+/// ```text
+/// Pattern              -> ValueSet
+/// ─────────────────────────────────────
+/// `_` or `other`       -> All (everything)
+/// `s: Success`         -> OfType("Success")
+/// `Status.Active`      -> EnumVariant("Status", "Active")
+/// `42`                 -> Literal(Int(42))
+/// `200 | 201`          -> Union([Literal(200), Literal(201)])
+/// `x: int if x > 0`    -> Empty (guards don't guarantee coverage)
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ValueSet {
+    /// Matches ALL possible values.
+    ///
+    /// This is the catch-all case: patterns like `_`, `other`, or any
+    /// untyped binding. It covers everything remaining.
+    All,
+
+    /// Matches all values of a named type.
+    ///
+    /// For **finite types** (enums, bool), this can be expanded into
+    /// the specific values. For **infinite types** (int, string, classes),
+    /// this represents an abstract "all instances of T".
+    ///
+    /// # Examples
+    /// - `s: Success` -> `OfType("Success")` (infinite: all Success instances)
+    /// - `b: bool` -> `OfType("bool")` -> expands to `[true, false]`
+    /// - `s: Status` -> `OfType("Status")` -> expands to variants
+    OfType(Name),
+
+    /// Matches exactly one enum variant value.
+    ///
+    /// # Example
+    /// `Status.Active` matches only the value `Status.Active`, not
+    /// `Status.Inactive` or any other value.
+    EnumVariant { enum_name: Name, variant_name: Name },
+
+    /// Matches exactly one literal value.
+    ///
+    /// # Examples
+    /// - `42` -> `Literal(Int(42))`
+    /// - `"hello"` -> `Literal(String("hello"))`
+    /// - `true` -> `Literal(Bool(true))`
+    /// - `null` -> `Literal(Null)`
+    Literal(Literal),
```

Comment:
Here I think it would make more sense to use the term "Singleton" for these two as opposed to using it in the general `Ty` struct. The meaning of singleton is important in the code that checks exhaustiveness with value sets.

Though this can stay as is, looks pretty clear to me without having to read term definitions.